python/tests/operations/join.py

```diff
@@ -0,0 +1,275 @@
+            join([self.ps_first_block, self.ps_first_block_extra_grad], axis="samples")
+        self.assertIn("gradients", str(err.exception))
+
+
+class TestJoinLabels(unittest.TestCase):
+    """Test edge cases of label joining."""
+
+    def setUp(self):
+        self.sample_labels = Labels(names=["prop"], values=np.arange(2).reshape(-1, 1))
+        self.keys = Labels(names=["prop"], values=np.arange(1).reshape(-1, 1))
+
+    def test_same_names_same_values(self):
+        """Test Label joining using labels with same names but same values."""
+
+        names = ("structure", "prop_1")
+        property_labels = Labels(names, np.vstack([np.arange(5), np.arange(5)]).T)
+
+        block = TensorBlock(
+            values=np.zeros([2, 5]),
+            samples=self.sample_labels,
+            components=[],
+            properties=property_labels,
+        )
+
+        ts = TensorMap(self.keys, [block])
+        joined_ts = join([ts, ts], axis="properties")
+
+        joined_labels = joined_ts.block(0).properties
+
+        self.assertEqual(joined_labels.names, ("tensor",) + names)
+
+        ref = np.array(
+            [
+                [0, 0, 0],
+                [0, 1, 1],
+                [0, 2, 2],
+                [0, 3, 3],
+                [0, 4, 4],
+                [1, 0, 0],
+                [1, 1, 1],
+                [1, 2, 2],
+                [1, 3, 3],
+                [1, 4, 4],
+            ]
+        )
+
+        self.assertTrue(np.equal(joined_labels.tolist(), ref).all())
+
+    def test_same_names_unique_values(self):
+        """Test Label joining using labels with same names and unique values."""
+        names = ("structure", "prop_1")
+        property_labels_1 = Labels(names, np.vstack([np.arange(5), np.arange(5)]).T)
+        property_labels_2 = Labels(
+            names, np.vstack([np.arange(5, 10), np.arange(5, 10)]).T
+        )
+
+        block_1 = TensorBlock(
+            values=np.zeros([2, 5]),
+            samples=self.sample_labels,
+            components=[],
+            properties=property_labels_1,
+        )
+
+        block_2 = TensorBlock(
+            values=np.zeros([2, 5]),
+            samples=self.sample_labels,
+            components=[],
+            properties=property_labels_2,
+        )
+
+        joined_ts = join(
+            [TensorMap(self.keys, [block_1]), TensorMap(self.keys, [block_2])],
+            axis="properties",
+        )
+
+        joined_labels = joined_ts.block(0).properties
+
+        self.assertEqual(joined_labels.names, ("structure", "prop_1"))
+        self.assertTrue(
+            np.equal(
+                joined_labels.tolist(), np.vstack([np.arange(10), np.arange(10)]).T
+            ).all()
+        )
+
+    def test_different_names(self):
+        """Test Label joining using labens with different names."""
+        values = np.vstack([np.arange(5), np.arange(5)]).T
+        property_labels_1 = Labels(("structure", "prop_1"), values)
```

Comment:
```suggestion
        property_labels_1 = Labels(("structure", "prop_1"), -values)
```

this way we check that the new values for `"property"` are actually re-created